### PR TITLE
Fixed #33651 -- Added support for prefetching GenericForeignKey.

### PR DESCRIPTION
**Pull request by clement-escolano**
Modified the prefetch queryset function in generic foreign key field to support custom queryset of different content types.

### CODE REVIEW

File: django/contrib/contenttypes/prefetch.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                or (
+                    hasattr(queryset, "_iterable_class")
+                    and not issubclass(queryset._iterable_class, ModelIterable)
+                )
+            ):
+                raise ValueError(
+                    "Prefetch querysets cannot use raw(), values(), and values_list()."
+                )
+        self.querysets = querysets
+        super(GenericPrefetch, self).__init__(lookup, to_attr=to_attr)
```
Comment (auvipy) [inline django/contrib/contenttypes/prefetch.py:19]:
```suggestion
        super().__init__(lookup, to_attr=to_attr)
```

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by smithdc1** [comment]
@clement-escolano I updated the flags on the ticket so it's now in the review queue. 

https://code.djangoproject.com/ticket/33651#comment:24

File: django/contrib/contenttypes/fields.py [hunk 2]
```diff
@@ -163,20 +165,45 @@ def _check_content_type_field(self):
...
             raise Exception("Impossible arguments to GFK.get_content_type!")
 
     def get_prefetch_queryset(self, instances, queryset=None):
-        if queryset is not None:
-            raise ValueError("Custom queryset can't be used for this lookup.")
+        warnings.warn(
+            "Calling get_prefetch_queryset() is deprecated. "
+            "Use get_prefetch_querysets() instead.",
+            RemovedInDjango60Warning,
+        )
```
Comment (felixxm) [inline django/contrib/contenttypes/fields.py:188]: This deprecation should be documented and tested, check out [docs](https://docs.djangoproject.com/en/4.2/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature).

File: django/db/models/query.py [hunk 3]
```diff
@@ -2225,8 +2226,22 @@ def get_current_to_attr(self, level):
         return to_attr, as_attr
 
     def get_current_queryset(self, level):
-        if self.get_current_prefetch_to(level) == self.prefetch_to:
-            return self.queryset
+        warnings.warn(
+            "Calling get_current_queryset() is deprecated. "
+            "Use get_current_querysets() instead.",
+            RemovedInDjango60Warning,
+        )
```
Comment (felixxm) [inline django/db/models/query.py:2233]: This deprecation should be documented and tested, check out [docs](https://docs.djangoproject.com/en/4.2/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature).

File: django/db/models/fields/related_descriptors.py [hunk 4]
```diff
@@ -427,18 +438,29 @@ def get_queryset(self, **hints):
...
         rel_obj_attr = self.related.field.get_local_related_value
         instance_attr = self.related.field.get_foreign_related_value
         instances_dict = {instance_attr(inst): inst for inst in instances}
         query = {"%s__in" % self.related.field.name: instances}
+
         queryset = queryset.filter(**query)
 
         # Since we're going to assign directly in the cache,
         # we must manage the reverse relation cache manually.
+
```
Comment (felixxm) [inline django/db/models/fields/related_descriptors.py:463]: Please revert unrelated blank lines.

File: django/contrib/contenttypes/fields.py [hunk 5]
```diff
@@ -616,9 +648,17 @@ def get_queryset(self):
...
+                "Use get_prefetch_querysets() instead.",
+                RemovedInDjango60Warning,
+            )
             if queryset is None:
-                queryset = super().get_queryset()
+                return self.get_prefetch_querysets(instances)
+            return self.get_prefetch_querysets(instances, [queryset])
 
+        def get_prefetch_querysets(self, instances, querysets=None):
+            queryset = querysets[0] if querysets is not None else super().get_queryset()
```
Comment (felixxm) [inline django/contrib/contenttypes/fields.py:661]: I would raise a `ValueError` when `len(queryset) > 1`.

File: docs/ref/contrib/contenttypes.txt [hunk 6]
```diff
@@ -590,3 +590,32 @@ information.
...
     layouts, respectively.
+
+.. module:: django.contrib.contenttypes.prefetch
+
+.. versionadded:: 5.0
+
+``GenericPrefetch()``
+
+-----------------------------
+
```
Comment (felixxm) [inline docs/ref/contrib/contenttypes.txt:601]:
```suggestion
``GenericPrefetch()``
---------------------

.. versionadded:: 5.0
```

File: docs/ref/contrib/contenttypes.txt [hunk 7]
```diff
@@ -590,3 +590,32 @@ information.
...
+.. class:: GenericPrefetch(lookup, querysets=None, to_attr=None)
+
+This lookup is similar to ``Prefetch()`` and it should only be used on
+``GenericForeignKey``. The ``querysets`` argument accepts a
+list of ``Queryset`` objects of different models. This is useful when we
+require each content type to prefetch their own queryset separately.
+
+    >>> bookmark = Bookmark.objects.create(url="https://www.djangoproject.com/")
```
Comment (felixxm) [inline docs/ref/contrib/contenttypes.txt:609]:
```suggestion
.. code-block:: pycon

    >>> bookmark = Bookmark.objects.create(url="https://www.djangoproject.com/")
```

**Review by felixxm — COMMENTED** [review]
@clement-escolano Thanks :+1: I left initial comments.

File: django/contrib/contenttypes/fields.py (hunk 2 above)
Comment (felixxm) [reply django/contrib/contenttypes/fields.py:188]: > should I add a section on the documentation about it?

No.

> In the release notes maybe?

Deprecation should be mentioned in release notes `5.0.txt` and `docs/internals/deprecation.txt`.

File: django/contrib/contenttypes/fields.py (hunk 5 above)
Comment (clement-escolano) [reply django/contrib/contenttypes/fields.py:661]: The behaviour of `get_prefetch_queryset` is duplicated 5 times and adding the length check would be also duplicated. Does it make sense to create a util method somewhere or is duplicating this behaviour (which will be removed anyway) fine?

File: django/contrib/contenttypes/fields.py (hunk 2 above)
Comment (clement-escolano) [reply django/contrib/contenttypes/fields.py:188]: Thank you! (I just noticed the what you said was written at the end of the deprecation doc, sorry)

File: django/contrib/contenttypes/fields.py (hunk 5 above)
Comment (clement-escolano) [reply django/contrib/contenttypes/fields.py:661]: The `get_prefetch_queryset` will be removed no? (This is the method that is marked as deprecated and that is also duplicated everywhere)

Comment (felixxm) [reply django/contrib/contenttypes/fields.py:661]: > The `get_prefetch_queryset` will be removed no? (This is the method that is marked as deprecated and that is also duplicated everywhere)

We are inside `get_prefetch_querysets()`, not in `get_prefetch_queryset()`.

Comment (clement-escolano) [reply django/contrib/contenttypes/fields.py:661]: OK 👍

**Comment by clement-escolano** [comment]
@felixxm Everything should be good now, thank you for the review. I added small commits for the feedbacks to make it easier to see the changes compared to my initial commit. If you prefer me squashing all the commits, I can do this too

File: django/contrib/contenttypes/fields.py [hunk 8]
```diff
@@ -616,9 +648,22 @@ def get_queryset(self):
...
-                queryset = super().get_queryset()
-
+                return self.get_prefetch_querysets(instances)
+            return self.get_prefetch_querysets(instances, [queryset])
+
+        def get_prefetch_querysets(self, instances, querysets=None):
+            if querysets is not None and len(querysets) != 1:
+                raise ValueError(
+                    "querysets argument of get_prefetch_querysets should be of length 1"
+                    f". Received an array of length {len(querysets)}"
```
Comment (felixxm) [inline django/contrib/contenttypes/fields.py:664]: The second sentence seems unnecessary. Also, f-strings should use only plain variable and property access. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: django/contrib/contenttypes/fields.py (hunk 2 above)
Comment (felixxm) [reply django/contrib/contenttypes/fields.py:188]: This warning is still untested.

File: django/contrib/contenttypes/fields.py [hunk 9]
```diff
@@ -163,20 +165,45 @@ def _check_content_type_field(self):
...
+        if querysets is not None:
+            for queryset in querysets:
+                ct_id = self.get_content_type(
+                    model=queryset.query.model, using=queryset.db
+                ).pk
+                if ct_id in custom_queryset_dict:
+                    raise ValueError(
+                        "Multiple querysets on same model found. "
+                        "Only one queryset on each model is allowed."
+                    )
```
Comment (felixxm) [inline django/contrib/contenttypes/fields.py:205]: As far as I'm aware, this is untested.

File: django/contrib/contenttypes/fields.py [hunk 10]
```diff
@@ -616,9 +648,22 @@ def get_queryset(self):
...
+            )
             if queryset is None:
-                queryset = super().get_queryset()
-
+                return self.get_prefetch_querysets(instances)
+            return self.get_prefetch_querysets(instances, [queryset])
+
+        def get_prefetch_querysets(self, instances, querysets=None):
+            if querysets is not None and len(querysets) != 1:
+                raise ValueError(
```
Comment (felixxm) [inline django/contrib/contenttypes/fields.py:662]: This is untested.

**Comment by clement-escolano** [comment]
@felixxm I added the missing tests and changed the Exception message. Is it OK now?

File: django/db/models/query.py [hunk 11]
```diff
@@ -2462,8 +2477,8 @@ def has_to_attr_attribute(instance):
         attr_found = True
         if rel_obj_descriptor:
             # singly related object, descriptor object has the
-            # get_prefetch_queryset() method.
-            if hasattr(rel_obj_descriptor, "get_prefetch_queryset"):
+            # get_prefetch_querysets() method.
+            if hasattr(rel_obj_descriptor, "get_prefetch_querysets"):
```
Comment (felixxm) [inline django/db/models/query.py:2481]: I'm afraid that to make it fully backward compatible we should check existence of both methods `get_prefetch_querysets` and `get_prefetch_queryset`, e.g.
```python
has_prefetch_qs = (
    hasattr(rel_obj_descriptor, "get_prefetch_querysets")
    or hasattr(rel_obj_descriptor, "get_prefetch_queryset")  # RemovedInDjango60Warning
)
```

File: django/db/models/query.py [hunk 12]
```diff
@@ -2517,7 +2532,9 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
         single,
         cache_name,
         is_descriptor,
-    ) = prefetcher.get_prefetch_queryset(instances, lookup.get_current_queryset(level))
+    ) = prefetcher.get_prefetch_querysets(
+        instances, lookup.get_current_querysets(level)
+    )
```
Comment (felixxm) [inline django/db/models/query.py:2537]: During the deprecation period, this should fallback to `get_prefetch_queryset()` when `get_prefetch_querysets()` is not implemented.

File: django/db/models/query.py (hunk 11 above)
Comment (clement-escolano) [reply django/db/models/query.py]: I changed the condition. It looks worse than on your example but I could not convince black to keep the two conditions on different lines with the comment on the second line.

I assume this should be tested too but I cannot figure out how to to id. Should I create a new field type with a new related manager that do not have the `get_prefetch_querysets` question?

**Review by felixxm — COMMENTED** [review]
I squashed, rebased, and pushed small edit to docs, will continue tomorrow.

**Review by felixxm — COMMENTED** [review]
I added tests for all new exceptions and deprecation warnings.

**Review by felixxm — COMMENTED** [review]
@clement-escolano Thanks :+1: Welcome aboard :boat:

**Review by auvipy — APPROVED** [review]
